api/turing/api/deployment_controller.go

```diff
@@ -231,6 +235,65 @@ func (c RouterDeploymentController) deployRouterVersion(
 		}
 	}
 
+	// Retrieve pyfunc ensembler if pyfunc ensembler is specified
+	if routerVersion.Ensembler != nil && routerVersion.Ensembler.Type == models.EnsemblerPyFuncType {
```

Comment:
Hmm yeah that sounds like a nice way to do it! It seems nicer to hide the `EnsemblerServiceBuilder` away in a `DeploymentService` struct instead of letting it sit at the `AppContext` level. The only odd part is probably where `DeployRouterVersion` would appear to be fiddling a bit with ensembler configuration instead of only managing k8s resources, but I guess that's a tiny detail we can overlook 😅

Reply:
Now that I've tried refactoring the stuff a little, it appears that it might be a bit complex trying to move `EnsemblerServiceBuilder` into `DeploymentService`... since this makes the service package where `DeploymentService` is from dependent on the imagebuilder package, which itself is dependent on the service package.. creating a cyclical dependency.

Reply:
Haha. Seems the only usage of the `service` package in the `imagebuilder` is a `service.EnsemblerFolder` from the `createKanikoJob` method, which IMO should not be directly referenced here in the first place (a generic image builder's method shouldn't need to know about the ensembler service). So, I think we can safely break this dependency and pass the `service.EnsemblerFolder` constant in as an argument to the imagebuilder's BuildImage().